expression/builtin_compare.go

```diff
@@ -385,14 +385,18 @@ func ResolveType4Between(args [3]Expression) types.EvalType {
 
 	hasTemporal := false
 	if cmpTp == types.ETString {
```

Comment:
We need to add some test cases in `expression/typeinfer_test.go:TestInferType`

Reply:
sorry for the later reply. @XuHuaiyu 
`ResolveType4Between` is used to rewrite the between query conditions, and the result is stored in the `allConds` field of the `DataSource`, which is private and cannot be obtained by `testInferTypeSuite`. Maybe we can add some similar test cases in `expression_rewriter_test.go`.